drivers/sensor/ina219/ina219.c

```diff
@@ -0,0 +1,358 @@
+/*
+ * Copyright (c) 2021 Leonard Pollak
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_ina219
+
+#include <device.h>
+#include <drivers/i2c.h>
+#include <kernel.h>
+#include <drivers/sensor.h>
+#include <sys/__assert.h>
+#include <logging/log.h>
+#include <sys/byteorder.h>
+
+#include <drivers/sensor/ina219.h>
+#include "ina219.h"
+
+LOG_MODULE_REGISTER(INA219, CONFIG_SENSOR_LOG_LEVEL);
+
+static int ina219_reg_read(const struct device *dev,
+		uint8_t reg_addr,
+		uint16_t *reg_data)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint8_t rx_buf[2];
+	int rc;
+
+	rc = i2c_write_read(cfg->bus, cfg->i2c_addr,
+			&reg_addr, sizeof(reg_addr),
+			rx_buf, sizeof(rx_buf));
+
+	*reg_data = sys_get_be16(rx_buf);
+
+	return rc;
+}
+
+static int ina219_reg_write(const struct device *dev,
+		uint8_t addr,
+		uint16_t reg_data)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint8_t tx_buf[3];
+
+	tx_buf[0] = addr;
+	sys_put_be16(reg_data, &tx_buf[1]);
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf), cfg->i2c_addr);
+}
+
+static int ina219_reg_field_update(const struct device *dev,
+		uint8_t addr,
+		uint16_t mask,
+		uint16_t field)
+{
+	uint16_t reg_data;
+
+	if (ina219_reg_read(dev, addr, &reg_data)) {
+		return -EIO;
+	}
+
+	reg_data = (reg_data & ~mask) | field;
+
+	return ina219_reg_write(dev, addr, reg_data);
+}
+
+static int ina219_set_msr_delay(const struct device *dev)
+{
+	const struct ina219_config *cfg = dev->config;
+	struct ina219_data *data = dev->data;
+
+	data->msr_delay =  ina219_conv_delay(cfg->badc) +
+		ina219_conv_delay(cfg->sadc);
+	return 0;
+}
+
+static int ina219_set_config(const struct device *dev)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint16_t reg_data;
+
+	reg_data = (cfg->brng & INA219_BRNG_MASK) << INA219_BRNG_SHIFT |
+		(cfg->pg & INA219_PG_MASK) << INA219_PG_SHIFT |
+		(cfg->badc & INA219_ADC_MASK) << INA219_BADC_SHIFT |
+		(cfg->sadc & INA219_ADC_MASK) << INA219_SADC_SHIFT |
+		(cfg->mode & INA219_MODE_NORMAL);
+
+
+	return ina219_reg_write(dev, INA219_REG_CONF, reg_data);
+}
+
+static int ina219_set_calib(const struct device *dev)
+{
+	const struct ina219_config *cfg = dev->config;
+	uint16_t cal;
+
+	cal = INA219_SCALING_FACTOR / ((cfg->r_shunt) * (cfg->current_lsb));
+
+	return ina219_reg_write(dev, INA219_REG_CALIB, cal);
+}
+
+
+static int ina219_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct ina219_data *data = dev->data;
+	int rc;
+	uint16_t status;
+	uint16_t tmp;
+
+	__ASSERT_NO_MSG(chan == SENSOR_CHAN_ALL);
```

Comment:
I'd return -ENOTSUP if given channel is not supported.

Reply:
i have updated the ina219_sample_fetch function to also be able to only fetch a single (private) channel.
not sure if it good the way i did it...